packages/test/src/test-worker-versioning.ts

```diff
@@ -0,0 +1,181 @@
+      taskQueue,
+      workflowId: wf2Id,
+    });
+    await wf1.signal(unblockSignal);
+
+    const worker2 = await Worker.create({
+      workflowsPath: require.resolve('./workflows'),
+      activities,
+      taskQueue,
+      showStackTraceSources: true,
+      buildId: '2.0',
+      useWorkerVersioning: true,
+    });
+    const worker2Prom = worker2.run();
+    worker2Prom.catch((err) => {
+      t.fail('Worker 2.0 run error: ' + JSON.stringify(err));
+    });
+
+    await wf2.signal(unblockSignal);
+
+    await wf1.result();
+    await wf2.result();
+
+    worker1.shutdown();
+    worker2.shutdown();
+    await worker1Prom;
+    await worker2Prom;
+    t.pass();
+  });
+
+  test('Worker versioning client updates', withSkipper, async (t) => {
+    const taskQueue = 'worker-versioning-client-updates-' + uuid4();
+    const conn = t.context.client;
+
+    await conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+      operation: 'newIdInNewDefaultSet',
+      buildId: '1.0',
+    });
+    let resp = await conn.taskQueue.getWorkerBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId(), '1.0');
+
+    await conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+      operation: 'newCompatibleVersion',
+      buildId: '1.1',
+      existingCompatibleBuildId: '1.0',
+    });
+    resp = await conn.taskQueue.getWorkerBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId(), '1.1');
+
+    // Target nonexistent build ID
+    await t.throwsAsync(
+      conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+        operation: 'newCompatibleVersion',
+        buildId: '1.2',
+        existingCompatibleBuildId: 'amnotreal',
+      }),
+      { message: /amnotreal not found/ }
+    );
+
+    await conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+      operation: 'promoteBuildIdWithinSet',
+      buildId: '1.0',
+    });
+    resp = await conn.taskQueue.getWorkerBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId(), '1.0');
+
+    await conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+      operation: 'newIdInNewDefaultSet',
+      buildId: '2.0',
+    });
+    resp = await conn.taskQueue.getWorkerBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId(), '2.0');
+
+    await conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+      operation: 'promoteSetByBuildId',
+      buildId: '1.0',
+    });
+    resp = await conn.taskQueue.getWorkerBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId(), '1.0');
+
+    await conn.taskQueue.updateWorkerBuildIdCompatibility(taskQueue, {
+      operation: 'mergeSets',
+      primaryBuildId: '2.0',
+      secondaryBuildId: '1.0',
+    });
+    resp = await conn.taskQueue.getWorkerBuildIdCompatability(taskQueue);
+    assert.equal(resp?.defaultBuildId(), '2.0');
+
+    const reachResp = await conn.taskQueue.getBuildIdReachability({ buildIds: ['2.0', '1.0', '1.1'] });
```

Comment:
This whole part should run in an "eventually" block (retry for a while) because of eventual consistency.